### PR TITLE
Enable waiting for web console creation

### PR DESCRIPTION
**Pull request by lyuboasenov**

### CODE REVIEW

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs [hunk 1]
```diff
@@ -498,95 +508,31 @@ public class K8sRunspaceProvider : IRunspaceProvider {
 
       public IRunspaceInfo WaitCreateCompletion(IRunspaceInfo runspaceInfo) {
          var result = runspaceInfo;
```
Comment (atanasa) [inline src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:510]: Remove the result variable and use return statements. The code, as it is, has too many code branches to follow.

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs [hunk 2]
```diff
@@ -615,6 +561,202 @@ public class K8sRunspaceProvider : IRunspaceProvider {
          return result;
       }
 
+      public IWebConsoleInfo WaitCreateCompletion(IWebConsoleInfo webConsoleInfo) {
+         // Wait for the console pod to get ready
+         IWebConsoleInfo result = webConsoleInfo;
```
Comment (atanasa) [inline src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:566]: Again, this pattern of using temporal values in the result makes this code hard to understand. Remove the result variable

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs [hunk 3]
```diff
@@ -615,6 +561,202 @@ public class K8sRunspaceProvider : IRunspaceProvider {
...
+            pod = _client.CoreV1.ReadNamespacedPod(name, _namespace);
+         } catch (Exception exc) {
+            LogException(exc);
+            throw new RunspaceProviderException(
+                  string.Format(
+                     Resources.K8sRunspaceProvider_WaitCreateComplation_PodNotFound, name),
+                  exc);
+         }
+
+         if (pod != null) {
```
Comment (atanasa) [inline src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:700]: When is pod null and why not handle the null case instead?

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs [hunk 4]
```diff
@@ -615,6 +561,202 @@ public class K8sRunspaceProvider : IRunspaceProvider {
...
+               if (retryCount >= maxRetryCount) {
+                  // Timeout
+                  result = new K8sWebConsoleInfo {
+                     Id = result.Id,
+                     CreationState = RunspaceCreationState.Error,
+                     CreationError = new RunspaceProviderException(Resources.K8sRunspaceProvider_WaitCreateComplition_TimeOut)
+                  };
+               } else {
+                  // Success, everything should be in place
+                  result = new K8sWebConsoleInfo {
```
[the remark this answers is not in the document]
Comment (lyuboasenov) [reply src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:668]: There is a timeout mechanism in place.

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs (hunk 3 above)
Comment (lyuboasenov) [reply src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:700]: fixed

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs (hunk 2 above)
Comment (lyuboasenov) [reply src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:566]: fixed

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs (hunk 1 above)
Comment (lyuboasenov) [reply src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs]: fixed

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs [hunk 5]
```diff
@@ -497,122 +507,267 @@ public class K8sRunspaceProvider : IRunspaceProvider {
...
 
       public IRunspaceInfo WaitCreateCompletion(IRunspaceInfo runspaceInfo) {
-         var result = runspaceInfo;
-         if (result != null && result.CreationState == RunspaceCreationState.Creating) {
-            V1Pod pod = null;
+         V1Pod pod = null;
+         if (runspaceInfo == null) {
+            return runspaceInfo;
+         }
+         if (runspaceInfo != null) {
```
Comment (atanasa) [inline src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:514]: This check looks redundant

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs [hunk 6]
```diff
@@ -497,122 +507,267 @@ public class K8sRunspaceProvider : IRunspaceProvider {
...
             }
          }
-         return result;
+      }
+
+      private static bool IsNginxReloadEventAfter(Corev1Event e, DateTime since) {
+         return e.LastTimestamp?.CompareTo(since) > 0 &&
+            e.Type.Equals("Normal", StringComparison.InvariantCultureIgnoreCase) &&
+            e.Reason.Equals("RELOAD", StringComparison.InvariantCultureIgnoreCase) &&
+            e.Message.Equals("NGINX reload triggered due to a change in configuration");
```
Comment (atanasa) [inline src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:692]: This `equals` seems fragile. Put in the backlog a task for creating a heartbeat API to replace this check here.

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs [hunk 7]
```diff
@@ -497,122 +507,267 @@ public class K8sRunspaceProvider : IRunspaceProvider {
...
+         //
+         // ImagePullBackOff - Container image pull failed, kubelet is backing off image pull
+         // ImageInspectError - Unable to inspect image
+         // ErrImagePull - General image pull error
+         // ErrImageNeverPull - Required Image is absent on host and PullPolicy is NeverPullImage
+         // RegistryUnavailable - Get http error when pulling image from registry
+         // InvalidImageName - Unable to parse the image name.
+
+         while (
+            pod != null &&
```
Comment (atanasa) [inline src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:746]: Put brackets to indicate the order of the operators && and ||

**Review by atanasa — APPROVED** [review]
Looks good to me

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs (hunk 7 above)
Comment (lyuboasenov) [reply src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:746]: fixed

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs (hunk 6 above)
Comment (lyuboasenov) [reply src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:692]: done

File: src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs (hunk 5 above)
Comment (lyuboasenov) [reply src/RunspaceProviders/Kubernetes/VMware.ScriptRuntimeService.K8sRunspaceProvider/K8sRunspaceProvider.cs:514]: fixed

**Pull request merged**